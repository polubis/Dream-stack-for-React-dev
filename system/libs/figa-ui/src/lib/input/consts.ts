const INPUT_VARIANTS = ['empty', 'filled', 'outlined'] as const;

const PREFIX_VARIANTS = [
  {
    key: 1,
    prefix: 'Mr.',
  },
  {
    key: 2,
    prefix: 'Mrs.',
  },
  {
    key: 3,
    prefix: 'Miss.',
  },
  {
    key: 4,
    prefix: 'Ms.',
  },
];

export { INPUT_VARIANTS, PREFIX_VARIANTS };
