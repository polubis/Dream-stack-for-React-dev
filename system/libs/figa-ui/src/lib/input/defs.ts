import type {
  DetailedHTMLProps,
  InputHTMLAttributes,
  ReactNode,
  TextareaHTMLAttributes,
} from 'react';
import type { INPUT_VARIANTS } from './consts';

type InputHTMLElementProps = DetailedHTMLProps<
  InputHTMLAttributes<HTMLInputElement>,
  HTMLInputElement
>;

type TextareaHTMLElementProps = DetailedHTMLProps<
  TextareaHTMLAttributes<HTMLTextAreaElement>,
  HTMLTextAreaElement
>;

type InputVariant = (typeof INPUT_VARIANTS)[number];

type ControlProps  = {
  className?: string;
  variant?: InputVariant;
  maxWidth?: string;
  disabled?: boolean;
  minWidth?: string;
  invalid?: boolean;
  loading?: boolean;
  children: ReactNode;
}

type BaseProps = Omit<ControlProps, 'children' | 'disabled' | 'className'>;

type InputProps = InputHTMLElementProps & BaseProps;

type TextareaProps = TextareaHTMLElementProps &
  BaseProps & { minHeight?: string; maxHeight?: string };

export type {
  InputProps,
  TextareaHTMLElementProps,
  InputHTMLElementProps,
  TextareaProps,
  ControlProps,
};
