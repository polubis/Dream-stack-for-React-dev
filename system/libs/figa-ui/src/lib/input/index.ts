export * from './defs';
export * from './input';
