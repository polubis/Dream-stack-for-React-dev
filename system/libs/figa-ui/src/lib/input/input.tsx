import type { InputProps } from './defs';

import c from 'classnames';

const Input = ({ className, ...inputProps }: InputProps) => {
  return (
    <div className={c('input', className)}>
      <input {...inputProps} />
    </div>
  );
};

export { Input };
