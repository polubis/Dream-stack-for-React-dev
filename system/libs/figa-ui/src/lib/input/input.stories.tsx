import type { Story, Meta } from '@storybook/react';

import { Input } from './input';

export default {
  component: Input,
  title: 'Input',
} as Meta;

const Template: Story = () => {
  return (
    <div
      style={{
        display: 'flex',
        flexFlow: 'column',
        gap: '20px',
      }}
    >
      <Input />
      <Input placeholder="Please write something..." />
      <div style={{ width: '220px' }}>
        <Input value="custom value" />
      </div>
    </div>
  );
};

export const Default = Template.bind({});
Default.args = {};
