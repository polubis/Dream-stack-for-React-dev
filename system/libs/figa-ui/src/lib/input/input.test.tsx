import type { ChangeEvent } from 'react';

import { fireEvent, render, screen } from '@testing-library/react';

import { Input, Textarea } from './input';

describe('User is able to user control when', () => {
  const PLACEHOLDER = 'type...';

  describe('input', () => {
    it('[FRAGILE] renders with default setup', () => {
      const { asFragment } = render(<Input />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('allows to use placeholder', () => {
      render(<Input placeholder={PLACEHOLDER} />);

      screen.getByPlaceholderText(PLACEHOLDER);
    });

    it('renders loader if loading', () => {
      const { asFragment } = render(
        <Input placeholder={PLACEHOLDER} loading />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('makes input disabled', () => {
      const { asFragment } = render(
        <Input placeholder={PLACEHOLDER} disabled />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('renders error icon if invalid', () => {
      const { asFragment } = render(
        <Input placeholder={PLACEHOLDER} invalid />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('assigns classes when additional options passed', () => {
      const { asFragment } = render(
        <Input
          placeholder={PLACEHOLDER}
          className="my-input"
          disabled
          loading
          invalid
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('allows to use native input properties', () => {
      const onChangeSpy = jest.fn();

      render(<Input placeholder={PLACEHOLDER} onChange={onChangeSpy} />);

      fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER), {
        target: { value: 'value' },
      } as ChangeEvent<HTMLInputElement>);

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
    });

    it('assigns classes', () => {
      render(<Input className="my-class" placeholder={PLACEHOLDER} />);

      const element = screen.getByPlaceholderText(PLACEHOLDER);

      expect(element.parentElement?.className).toBe('filled input my-class');
    });
  });

  describe('textarea', () => {
    it('[FRAGILE] renders with default setup', () => {
      const { asFragment } = render(<Textarea />);

      expect(asFragment()).toMatchSnapshot();
    });

    it('allows to use placeholder', () => {
      render(<Textarea placeholder={PLACEHOLDER} />);

      screen.getByPlaceholderText(PLACEHOLDER);
    });

    it('allows to set max height and min height', () => {
      const { asFragment } = render(
        <Textarea
          placeholder={PLACEHOLDER}
          maxHeight="300px"
          minHeight="200px"
          style={{
            maxHeight: '270px',
            minHeight: '190px',
          }}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('renders loader if loading', () => {
      const { asFragment } = render(
        <Textarea placeholder={PLACEHOLDER} loading />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('makes input disabled', () => {
      const { asFragment } = render(
        <Textarea placeholder={PLACEHOLDER} disabled />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('renders error icon if invalid', () => {
      const { asFragment } = render(
        <Textarea placeholder={PLACEHOLDER} invalid />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('assigns classes when additional options passed', () => {
      const { asFragment } = render(
        <Textarea
          placeholder={PLACEHOLDER}
          className="my-input"
          disabled
          loading
          invalid
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('allows to use native input properties', () => {
      const onChangeSpy = jest.fn();

      render(<Textarea placeholder={PLACEHOLDER} onChange={onChangeSpy} />);

      fireEvent.change(screen.getByPlaceholderText(PLACEHOLDER), {
        target: { value: 'value' },
      } as ChangeEvent<HTMLInputElement>);

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
    });

    it('assigns classes', () => {
      render(<Textarea className="my-class" placeholder={PLACEHOLDER} />);

      const element = screen.getByPlaceholderText(PLACEHOLDER);

      expect(element.parentElement?.className).toBe('filled textarea my-class');
    });
  });
});
