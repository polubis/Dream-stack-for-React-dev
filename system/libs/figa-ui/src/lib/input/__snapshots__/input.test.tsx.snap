// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User is able to user control when input [FRAGILE] renders with default setup 1`] = `
<DocumentFragment>
  <div
    class="filled input"
  >
    <input />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input assigns classes when additional options passed 1`] = `
<DocumentFragment>
  <div
    class="filled invalid disabled loading input my-input"
  >
    <div
      class="loader tiny input-loader"
    >
      <div
        class="loader-animation loader-1"
      >
        <div>
          <div>
            <div>
              <div>
                <div>
                  <div />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <input
      disabled=""
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input makes input disabled 1`] = `
<DocumentFragment>
  <div
    class="filled disabled input"
  >
    <input
      disabled=""
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input renders error icon if invalid 1`] = `
<DocumentFragment>
  <div
    class="filled invalid input"
  >
    <svg
      class="icon error-icon input-icon"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1.5 12C1.5 14.0767 2.11581 16.1068 3.26957 17.8335C4.42332 19.5602 6.0632 20.906 7.98182 21.7007C9.90045 22.4955 12.0116 22.7034 14.0484 22.2982C16.0852 21.8931 17.9562 20.8931 19.4246 19.4246C20.8931 17.9562 21.8931 16.0852 22.2982 14.0484C22.7034 12.0116 22.4955 9.90045 21.7007 7.98182C20.906 6.0632 19.5602 4.42332 17.8335 3.26957C16.1068 2.11581 14.0767 1.5 12 1.5C9.21523 1.5 6.54451 2.60625 4.57538 4.57538C2.60625 6.54451 1.5 9.21523 1.5 12ZM18.8625 17.8125L6.1875 5.1375C7.91298 3.70125 10.1129 2.96173 12.3556 3.0641C14.5983 3.16646 16.7218 4.10331 18.3092 5.69077C19.8967 7.27824 20.8335 9.40169 20.9359 11.6444C21.0383 13.8871 20.2988 16.087 18.8625 17.8125ZM6.18 18.87C4.36374 17.3276 3.23325 15.1279 3.03642 12.7533C2.83959 10.3787 3.59248 8.02286 5.13 6.2025L17.7975 18.87C16.1732 20.2425 14.1153 20.9955 11.9887 20.9955C9.86219 20.9955 7.80433 20.2425 6.18 18.87Z"
      />
    </svg>
    <input
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input renders loader if loading 1`] = `
<DocumentFragment>
  <div
    class="filled loading input"
  >
    <div
      class="loader tiny input-loader"
    >
      <div
        class="loader-animation loader-1"
      >
        <div>
          <div>
            <div>
              <div>
                <div>
                  <div />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <input
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input renders prefix and suffix 1`] = `
<DocumentFragment>
  <div
    class="filled input prefx suffx"
  >
    <div
      class="prefx-wrapper"
    >
      <div>
        A
      </div>
    </div>
    <input />
    <div
      class="suffx-wrapper"
    >
      <div>
        S
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input skips suffix when invalid 1`] = `
<DocumentFragment>
  <div
    class="filled invalid input prefx suffx"
  >
    <svg
      class="icon error-icon input-icon"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1.5 12C1.5 14.0767 2.11581 16.1068 3.26957 17.8335C4.42332 19.5602 6.0632 20.906 7.98182 21.7007C9.90045 22.4955 12.0116 22.7034 14.0484 22.2982C16.0852 21.8931 17.9562 20.8931 19.4246 19.4246C20.8931 17.9562 21.8931 16.0852 22.2982 14.0484C22.7034 12.0116 22.4955 9.90045 21.7007 7.98182C20.906 6.0632 19.5602 4.42332 17.8335 3.26957C16.1068 2.11581 14.0767 1.5 12 1.5C9.21523 1.5 6.54451 2.60625 4.57538 4.57538C2.60625 6.54451 1.5 9.21523 1.5 12ZM18.8625 17.8125L6.1875 5.1375C7.91298 3.70125 10.1129 2.96173 12.3556 3.0641C14.5983 3.16646 16.7218 4.10331 18.3092 5.69077C19.8967 7.27824 20.8335 9.40169 20.9359 11.6444C21.0383 13.8871 20.2988 16.087 18.8625 17.8125ZM6.18 18.87C4.36374 17.3276 3.23325 15.1279 3.03642 12.7533C2.83959 10.3787 3.59248 8.02286 5.13 6.2025L17.7975 18.87C16.1732 20.2425 14.1153 20.9955 11.9887 20.9955C9.86219 20.9955 7.80433 20.2425 6.18 18.87Z"
      />
    </svg>
    <div
      class="prefx-wrapper"
    >
      <div>
        A
      </div>
    </div>
    <input />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when input skips suffix when loading 1`] = `
<DocumentFragment>
  <div
    class="filled loading input prefx suffx"
  >
    <div
      class="loader tiny input-loader"
    >
      <div
        class="loader-animation loader-1"
      >
        <div>
          <div>
            <div>
              <div>
                <div>
                  <div />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="prefx-wrapper"
    >
      <div>
        A
      </div>
    </div>
    <input />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when textarea [FRAGILE] allows to set style 1`] = `
<DocumentFragment>
  <div
    class="filled textarea"
  >
    <textarea
      placeholder="type..."
      style="max-height: 270px; min-height: 190px;"
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when textarea [FRAGILE] renders with default setup 1`] = `
<DocumentFragment>
  <div
    class="filled textarea"
  >
    <textarea />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when textarea assigns classes when additional options passed 1`] = `
<DocumentFragment>
  <div
    class="filled invalid disabled loading textarea my-input"
  >
    <div
      class="loader tiny input-loader"
    >
      <div
        class="loader-animation loader-1"
      >
        <div>
          <div>
            <div>
              <div>
                <div>
                  <div />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <textarea
      disabled=""
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when textarea makes input disabled 1`] = `
<DocumentFragment>
  <div
    class="filled disabled textarea"
  >
    <textarea
      disabled=""
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when textarea renders error icon if invalid 1`] = `
<DocumentFragment>
  <div
    class="filled invalid textarea"
  >
    <svg
      class="icon error-icon input-icon"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1.5 12C1.5 14.0767 2.11581 16.1068 3.26957 17.8335C4.42332 19.5602 6.0632 20.906 7.98182 21.7007C9.90045 22.4955 12.0116 22.7034 14.0484 22.2982C16.0852 21.8931 17.9562 20.8931 19.4246 19.4246C20.8931 17.9562 21.8931 16.0852 22.2982 14.0484C22.7034 12.0116 22.4955 9.90045 21.7007 7.98182C20.906 6.0632 19.5602 4.42332 17.8335 3.26957C16.1068 2.11581 14.0767 1.5 12 1.5C9.21523 1.5 6.54451 2.60625 4.57538 4.57538C2.60625 6.54451 1.5 9.21523 1.5 12ZM18.8625 17.8125L6.1875 5.1375C7.91298 3.70125 10.1129 2.96173 12.3556 3.0641C14.5983 3.16646 16.7218 4.10331 18.3092 5.69077C19.8967 7.27824 20.8335 9.40169 20.9359 11.6444C21.0383 13.8871 20.2988 16.087 18.8625 17.8125ZM6.18 18.87C4.36374 17.3276 3.23325 15.1279 3.03642 12.7533C2.83959 10.3787 3.59248 8.02286 5.13 6.2025L17.7975 18.87C16.1732 20.2425 14.1153 20.9955 11.9887 20.9955C9.86219 20.9955 7.80433 20.2425 6.18 18.87Z"
      />
    </svg>
    <textarea
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;

exports[`User is able to user control when textarea renders loader if loading 1`] = `
<DocumentFragment>
  <div
    class="filled loading textarea"
  >
    <div
      class="loader tiny input-loader"
    >
      <div
        class="loader-animation loader-1"
      >
        <div>
          <div>
            <div>
              <div>
                <div>
                  <div />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <textarea
      placeholder="type..."
    />
  </div>
</DocumentFragment>
`;
